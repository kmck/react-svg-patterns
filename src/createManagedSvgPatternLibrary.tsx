import * as React from 'react';

import { PatternType } from './types';

import SvgPatternLibrary from './SvgPatternLibrary';
import SvgPatternManager from './SvgPatternManager';

interface Props {
  noSvgWrapper?: boolean;
}

export default function createManagedSvgPatternLibrary(
  patternManagerOrId?: string | SvgPatternManager
) {
  const patternManager =
    (patternManagerOrId as any) instanceof SvgPatternManager
      ? (patternManagerOrId as SvgPatternManager)
      : new SvgPatternManager(patternManagerOrId as string);
  const ManagedSvgPatternLibrary = (props: Props) => {
    const [patterns, setPatterns] = React.useState(patternManager.getPatterns());

    React.useEffect(() => {
      setPatterns(patternManager.getPatterns());

      const unsubscribe = patternManager.subscribe(() => {
        setPatterns(patternManager.getPatterns());
      });

      return () => {
        unsubscribe();
      };
    }, []);

    return <SvgPatternLibrary patterns={patterns} {...props} />;
  };

  return {
    ManagedSvgPatternLibrary,
    getSvgPattern(key: string) {
      return patternManager.get(key);
    },
    registerSvgPattern(key: string, type: PatternType, params: Object) {
      return patternManager.get(patternManager.add(key, type, params));
    },
    useSvgPattern(key: string, type: PatternType, params: Object) {
      const [pattern, setPattern] = React.useState<string | undefined>(undefined);
      React.useLayoutEffect(() => {
        setPattern(patternManager.get(patternManager.add(key, type, params)));
      }, [key, type, params]);
      return pattern;
    },
  };
}
