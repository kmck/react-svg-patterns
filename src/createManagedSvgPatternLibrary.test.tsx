import * as React from 'react';
import renderer from 'react-test-renderer';

import createManagedSvgPatternLibrary from './createManagedSvgPatternLibrary';

type CreateManagedSvgPatternLibraryReturn = ReturnType<typeof createManagedSvgPatternLibrary>;

describe('createManagedSvgPatternLibrary', () => {
  let ManagedSvgPatternLibrary: CreateManagedSvgPatternLibraryReturn['ManagedSvgPatternLibrary'];
  let getSvgPattern: CreateManagedSvgPatternLibraryReturn['getSvgPattern'];
  let registerSvgPattern: CreateManagedSvgPatternLibraryReturn['registerSvgPattern'];
  let useSvgPattern: CreateManagedSvgPatternLibraryReturn['useSvgPattern'];

  beforeEach(() => {
    ({
      ManagedSvgPatternLibrary,
      getSvgPattern,
      registerSvgPattern,
      useSvgPattern,
    } = createManagedSvgPatternLibrary('testManager'));
  });

  describe('registerSvgPattern', () => {
    it('returns the identifier of the registered pattern', () => {
      const valueActual = registerSvgPattern('test', 'linear', { from: '#f00', to: '#00f' });
      const valueExpected = 'url(#testManager-linear__test)';
      expect(valueActual).toBe(valueExpected);
    });

    it('converts whitespace to dashes', () => {
      const valueActual = registerSvgPattern('extra space', 'linear', { from: '#f00', to: '#00f' });
      const valueExpected = 'url(#testManager-linear__extra-space)';
      expect(valueActual).toBe(valueExpected);
    });
  });

  describe('getSvgPattern', () => {
    it('returns the identifier of a registered pattern', () => {
      registerSvgPattern('test', 'linear', { from: '#f00', to: '#00f' });
      const valueActual = getSvgPattern('test');
      expect(valueActual).toBe('url(#testManager-linear__test)');
    });

    it('returns null when the key is not registered', () => {
      expect(getSvgPattern('miss')).toBeUndefined();
    });

    it('matches the identifier returned by registerSvgPattern', () => {
      const valueExpected = registerSvgPattern('my cool gradient', 'linear', { from: '#f00', to: '#00f' });
      const valueActual = getSvgPattern('my cool gradient');
      expect(valueActual).toBe(valueExpected);
    });
  });

  describe('ManagedSvgPatternLibrary', () => {
    it('renders with no fills', () => {
      const tree = renderer.create(
        <ManagedSvgPatternLibrary />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders with all fill types', () => {
      registerSvgPattern('test1', 'linear', { from: '#f00', to: '#00f' });
      registerSvgPattern('test2', 'radial', { from: '#f00', to: '#00f' });
      registerSvgPattern('test3', 'angular', { from: '#f00', to: '#00f', slices: 4 });
      registerSvgPattern('test4', 'image', { src: 'https://via.placeholder.com/150' });
      registerSvgPattern('test5', 'custom', { children: () => <rect x={0} y={0} width={1} height={1} fill="#f00" /> });
      const tree = renderer.create(
        <ManagedSvgPatternLibrary />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders without an SVG wrapper', () => {
      registerSvgPattern('test1', 'linear', { from: '#f00', to: '#00f' });
      registerSvgPattern('test2', 'radial', { from: '#f00', to: '#00f' });
      const tree = renderer.create(
        <ManagedSvgPatternLibrary noSvgWrapper />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('useSvgPattern', () => {
    it('renders with all fill types', () => {
      const FillTest = () => {
        const testFill1 = useSvgPattern('test1', 'linear', { from: '#f00', to: '#00f' });
        const testFill2 = useSvgPattern('test2', 'radial', { from: '#f00', to: '#00f' });
        const testFill3 = useSvgPattern('test3', 'angular', { from: '#f00', to: '#00f', slices: 4 });
        const testFill4 = useSvgPattern('test4', 'image', { src: 'https://via.placeholder.com/150' });
        const testFill5 = useSvgPattern('test5', 'custom', { children: () => <rect x={0} y={0} width={1} height={1} fill="#f00" /> });
        return (
          <svg
            width={100}
            height={100}
            viewBox="0 0 100 100"
          >
            <rect x={0} y={0} width={100} height={20} fill={testFill1} />
            <rect x={0} y={20} width={100} height={20} fill={testFill2} />
            <rect x={0} y={40} width={100} height={20} fill={testFill3} />
            <rect x={0} y={60} width={100} height={20} fill={testFill4} />
            <rect x={0} y={80} width={100} height={20} fill={testFill5} />
          </svg>
        );
      };
      const tree = renderer.create(
        <>
          <FillTest />
          <ManagedSvgPatternLibrary />
        </>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
