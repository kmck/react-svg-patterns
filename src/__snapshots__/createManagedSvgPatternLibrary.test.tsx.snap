// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createManagedSvgPatternLibrary ManagedSvgPatternLibrary renders with all fill types 1`] = `
<svg
  style={
    Object {
      "bottom": 0,
      "height": 1,
      "opacity": 0,
      "pointerEvents": "none",
      "position": "fixed",
      "right": 0,
      "width": 1,
      "zIndex": -1,
    }
  }
>
  <linearGradient
    id="testManager-linear__test1"
    spreadMethod="pad"
    x1={0.5}
    x2={0.5}
    y1={1}
    y2={0}
  >
    <stop
      offset="0%"
      stopColor="#f00"
    />
    <stop
      offset="100%"
      stopColor="#00f"
    />
  </linearGradient>
  <radialGradient
    cx={0.5}
    cy={0.5}
    fx={0.5}
    fy={0.5}
    id="testManager-radial__test2"
    r={0.5}
    spreadMethod="pad"
  >
    <stop
      offset="0%"
      stopColor="#f00"
    />
    <stop
      offset="100%"
      stopColor="#00f"
    />
  </radialGradient>
  <pattern
    height={1}
    id="testManager-angular__test3"
    patternContentUnits="objectBoundingBox"
    patternUnits="objectBoundingBox"
    width={1}
  >
    <polygon
      fill="#ff0000"
      points="0.5,0.5 0.5,-0.5 1.2071067811865475,-0.20710678118654757 1.5,0.49999999999999994 1.2071067811865475,1.2071067811865475"
    />
    <polygon
      fill="#aa0055"
      points="0.5,0.5 1.5,0.49999999999999994 1.2071067811865475,1.2071067811865475 0.5000000000000001,1.5 -0.20710678118654746,1.2071067811865477"
    />
    <polygon
      fill="#5500aa"
      points="0.5,0.5 0.5000000000000001,1.5 -0.20710678118654746,1.2071067811865477 -0.5,0.5000000000000002 -0.20710678118654768,-0.20710678118654735"
    />
    <polygon
      fill="#0000ff"
      points="0.5,0.5 -0.5,0.5000000000000002 -0.20710678118654768,-0.20710678118654735 0.4999999999999998,-0.5"
    />
  </pattern>
  <pattern
    height={1}
    id="testManager-image__test4"
    patternContentUnits="objectBoundingBox"
    patternUnits="objectBoundingBox"
    width={1}
  >
    <image
      height={1}
      width={1}
      x={0}
      xlinkHref="https://via.placeholder.com/150"
      y={0}
    />
  </pattern>
  <pattern
    height={1}
    id="testManager-custom__test5"
    patternContentUnits="objectBoundingBox"
    patternUnits="objectBoundingBox"
    width={1}
  >
    <rect
      fill="#f00"
      height={1}
      width={1}
      x={0}
      y={0}
    />
  </pattern>
</svg>
`;

exports[`createManagedSvgPatternLibrary ManagedSvgPatternLibrary renders with no fills 1`] = `
<svg
  style={
    Object {
      "bottom": 0,
      "height": 1,
      "opacity": 0,
      "pointerEvents": "none",
      "position": "fixed",
      "right": 0,
      "width": 1,
      "zIndex": -1,
    }
  }
/>
`;

exports[`createManagedSvgPatternLibrary ManagedSvgPatternLibrary renders without an SVG wrapper 1`] = `
Array [
  <linearGradient
    id="testManager-linear__test1"
    spreadMethod="pad"
    x1={0.5}
    x2={0.5}
    y1={1}
    y2={0}
  >
    <stop
      offset="0%"
      stopColor="#f00"
    />
    <stop
      offset="100%"
      stopColor="#00f"
    />
  </linearGradient>,
  <radialGradient
    cx={0.5}
    cy={0.5}
    fx={0.5}
    fy={0.5}
    id="testManager-radial__test2"
    r={0.5}
    spreadMethod="pad"
  >
    <stop
      offset="0%"
      stopColor="#f00"
    />
    <stop
      offset="100%"
      stopColor="#00f"
    />
  </radialGradient>,
]
`;

exports[`createManagedSvgPatternLibrary useSvgPattern renders with all fill types 1`] = `
Array [
  <svg
    height={100}
    viewBox="0 0 100 100"
    width={100}
  >
    <rect
      fill="url(#testManager-linear__test1)"
      height={20}
      width={100}
      x={0}
      y={0}
    />
    <rect
      fill="url(#testManager-radial__test2)"
      height={20}
      width={100}
      x={0}
      y={20}
    />
    <rect
      fill="url(#testManager-angular__test3)"
      height={20}
      width={100}
      x={0}
      y={40}
    />
    <rect
      fill="url(#testManager-image__test4)"
      height={20}
      width={100}
      x={0}
      y={60}
    />
    <rect
      fill="url(#testManager-custom__test5)"
      height={20}
      width={100}
      x={0}
      y={80}
    />
  </svg>,
  <svg
    style={
      Object {
        "bottom": 0,
        "height": 1,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "fixed",
        "right": 0,
        "width": 1,
        "zIndex": -1,
      }
    }
  >
    <linearGradient
      id="testManager-linear__test1"
      spreadMethod="pad"
      x1={0.5}
      x2={0.5}
      y1={1}
      y2={0}
    >
      <stop
        offset="0%"
        stopColor="#f00"
      />
      <stop
        offset="100%"
        stopColor="#00f"
      />
    </linearGradient>
    <radialGradient
      cx={0.5}
      cy={0.5}
      fx={0.5}
      fy={0.5}
      id="testManager-radial__test2"
      r={0.5}
      spreadMethod="pad"
    >
      <stop
        offset="0%"
        stopColor="#f00"
      />
      <stop
        offset="100%"
        stopColor="#00f"
      />
    </radialGradient>
    <pattern
      height={1}
      id="testManager-angular__test3"
      patternContentUnits="objectBoundingBox"
      patternUnits="objectBoundingBox"
      width={1}
    >
      <polygon
        fill="#ff0000"
        points="0.5,0.5 0.5,-0.5 1.2071067811865475,-0.20710678118654757 1.5,0.49999999999999994 1.2071067811865475,1.2071067811865475"
      />
      <polygon
        fill="#aa0055"
        points="0.5,0.5 1.5,0.49999999999999994 1.2071067811865475,1.2071067811865475 0.5000000000000001,1.5 -0.20710678118654746,1.2071067811865477"
      />
      <polygon
        fill="#5500aa"
        points="0.5,0.5 0.5000000000000001,1.5 -0.20710678118654746,1.2071067811865477 -0.5,0.5000000000000002 -0.20710678118654768,-0.20710678118654735"
      />
      <polygon
        fill="#0000ff"
        points="0.5,0.5 -0.5,0.5000000000000002 -0.20710678118654768,-0.20710678118654735 0.4999999999999998,-0.5"
      />
    </pattern>
    <pattern
      height={1}
      id="testManager-image__test4"
      patternContentUnits="objectBoundingBox"
      patternUnits="objectBoundingBox"
      width={1}
    >
      <image
        height={1}
        width={1}
        x={0}
        xlinkHref="https://via.placeholder.com/150"
        y={0}
      />
    </pattern>
    <pattern
      height={1}
      id="testManager-custom__test5"
      patternContentUnits="objectBoundingBox"
      patternUnits="objectBoundingBox"
      width={1}
    >
      <rect
        fill="#f00"
        height={1}
        width={1}
        x={0}
        y={0}
      />
    </pattern>
  </svg>,
]
`;
